quil-py/src/program/source_map.rs

```diff
@@ -0,0 +1,147 @@
+use pyo3::{types::PyInt, Py};
+use quil_rs::program::{
+    CalibrationExpansion, CalibrationSource, MaybeCalibrationExpansion,
+    ProgramCalibrationExpansion, ProgramCalibrationExpansionSourceMap, SourceMap, SourceMapEntry,
+};
+use rigetti_pyo3::{impl_repr, py_wrap_type, py_wrap_union_enum, pyo3::pymethods, PyWrapper};
+
+use crate::instruction::{PyCalibrationIdentifier, PyMeasureCalibrationIdentifier};
+
+use super::PyProgram;
+
+type CalibrationExpansionSourceMap = SourceMap<usize, CalibrationExpansion>;
+type CalibrationExpansionSourceMapEntry = SourceMapEntry<usize, CalibrationExpansion>;
+type ProgramCalibrationExpansionSourceMapEntry = SourceMapEntry<usize, MaybeCalibrationExpansion>;
+
+py_wrap_type! {
+    #[derive(Debug, PartialEq)]
+    PyCalibrationExpansion(CalibrationExpansion) as "CalibrationExpansion"
+}
+
+impl_repr!(PyCalibrationExpansion);
+
+#[pymethods]
+impl PyCalibrationExpansion {
+    pub fn calibration_used(&self) -> PyCalibrationSource {
+        self.as_inner().calibration_used().into()
+    }
+
+    // Reviewer: is there a better return type?
```

Comment:
I'm not against this implementation, but an option is to return a Python range. That might be slightly more Pythonic:

```rs
pub fn range(&self) -> (usize, usize) {
    let range = py.import("builtins")?.get("range")?;
    let args = PyTuple::new(py, &[self.as_inner().range().start, self.as_inner().range().end]);
    range.call1(args)
}
```
Return type aside, I wonder if it's possible to only consume the inner range a single time, rather that once for the start and once for the end.

Reply:
sure - TIL this trick with pyo3. It's compiling but I don't think I did it right; will fix it up and test on the next pass.

Reply:
got this fixed up and under test - looking good. please lmk what you think

Reply:
Looks good to me!